### PR TITLE
Python 3+ support

### PR DESCRIPTION
**Pull request by diranged**

### CODE REVIEW

**Comment by diranged** [comment]
@swaggy, take a peek? I'd like to try to use this in our circleci-scheduler project..

File: tornado_rest_client/utils.py [hunk 1]
```diff
@@ -66,9 +66,13 @@ def populate_with_tokens(string, tokens, left_wrapper='%', right_wrapper='%',
     # First things first, swap out all instances of %<str>% with any matching
     # token variables found. If no items are in the hash (none, empty hash,
     # etc), then skip this.
-    allowed_types = (str, unicode, bool, int, float)
+    try:
+        allowed_types = (str, unicode, bool, int, float)
```
Comment (mrterry) [inline tornado_rest_client/utils.py]: there is no `unicode` in python 3.  do you want to handle `bytes`?

**Review by mrterry — CHANGES_REQUESTED** [review]
I'd run this through futurize.  http://python-future.org/automatic_conversion.html  based on its transformations, you might want to add six or future as a dependency.

**Comment by Stovoy** [comment]
@diranged what's this got over aiohttp?

File: tornado_rest_client/utils.py (hunk 1 above)
Comment (diranged) [reply tornado_rest_client/utils.py]: Yeah I figured out the whole unicode thing - which is why this is in a try/except block. I don't think that we want or need to handle bytes, but I may be wrong.

**Comment by diranged** [comment]
@mrterry Re your comment, I'll look at what Mr Brown did in https://github.com/Nextdoor/tornado_rest_client/pull/6/files and pull out the appropriate bits..

**Comment by diranged** [comment]
@mrterry, ok this has been six'd... looks like it works in py2 and py3.. ive actually got it running in py3 with asyncio as well..

File: tornado_rest_client/api.py [hunk 2]
```diff
@@ -125,7 +125,7 @@ def wrapper(self, *args, **kwargs):
                     # pop it before searching.
                     default_exc = exc_conf.pop('', False)
                     log.debug('Searching through %s' % exc_conf)
-                    matched_exc = [exc for key, exc in exc_conf.items()
+                    matched_exc = [exc for key, exc in list(exc_conf.items())
```
Comment (mrterry) [inline tornado_rest_client/api.py:128]: FYI: `dict.items()` is an iterator in python3, but a list in python2.  the automatic conversion tools are conservative and slap an unnecessary list coercion in places like this.

File: tornado_rest_client/test/integration_api.py [hunk 3]
```diff
@@ -70,7 +73,7 @@ class IntegrationRestConsumer(testing.AsyncTestCase):
     def integration_base_get(self):
         httpbin = HTTPBinRestConsumer()
         ret = yield httpbin.http_get()
-        self.assertIn('DOCTYPE', ret)
+        self.assertIn(six.b('DOCTYPE'), ret)
```
Comment (mrterry) [inline tornado_rest_client/test/integration_api.py:76]: both python2 and python3 have byte/unicode string literals.  `b'this is a byte string'` `u'this is a unicode string'`

File: tornado_rest_client/utils.py [hunk 4]
```diff
@@ -66,13 +68,14 @@ def populate_with_tokens(string, tokens, left_wrapper='%', right_wrapper='%',
     # First things first, swap out all instances of %<str>% with any matching
     # token variables found. If no items are in the hash (none, empty hash,
     # etc), then skip this.
-    allowed_types = (str, unicode, bool, int, float)
+    allowed_types = six.string_types + (six.text_type, str, bool, int, float)
+
     if tokens:
-        for k, v in tokens.iteritems():
+        for k, v in six.iteritems(tokens):
```
Comment (mrterry) [inline tornado_rest_client/utils.py:74]: i find `six.iteritems` to be really ugly.  unless you know a dict to be very large, and really need `iteritems()` i just use `items()` and move on with life.  in python3, `items()` is lazy like `iteritems()` is in python2

File: tornado_rest_client/utils.py [hunk 5]
```diff
@@ -66,13 +68,14 @@ def populate_with_tokens(string, tokens, left_wrapper='%', right_wrapper='%',
...
 
             if type(v) not in allowed_types:
-                log.warning('Token %s=%s is not in allowed types: %s' % (
-                    k, v, allowed_types))
+                log.warning('Token %s=%s (%s)is not in allowed types: %s' % (
+                    k, v, type(v), allowed_types))
                 continue
 
             string = string.replace(
```
Comment (mrterry) [inline tornado_rest_client/utils.py:81]: if k is `bytes` this will turn weird

```
In [1]: '%s%s%s' % ('hi', b'bye', 'hi')
Out[1]: "hib'bye'hi"```

Comment (mrterry) [reply tornado_rest_client/utils.py:81]: ditto for `v`

```
In [2]: str(b'bye')
Out[2]: "b'bye'"```

Comment (mrterry) [reply tornado_rest_client/utils.py:81]: i'd suggest only working with unicode.  `bytes` are for shipping things over the wire and should be immediately be turned into text (unicode) as soon as received.

**Review by mrterry — APPROVED** [review]
This looks fine.  You will probably have some funny edge cases wrt `bytes`/`str` handling, but no reason to hold this up.

**Pull request closed without merging**